Invalid GoTemplate can crash ApplicationSet Controller

<!-- If you are trying to resolve an environment-specific issue or have a one-off question about the edge case that does not require a feature then please consider asking a question in argocd slack [channel](https://argoproj.github.io/community/join-slack). -->

Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

When using the interpolated generator, an invalid GoTemplate can cause a nil pointer dereference that causes a panic and crashes the ApplicationSet controller. 

**To Reproduce**

See the unit test `TestInterpolateGeneratorError/Error_templating`
```
=== RUN   TestInterpolateGeneratorError/Error_templating
ERRO[0003] error interpolating generator with other generator's parameter  error="failed to replace parameters in generator: failed to execute go template {{ index .rmap (default .override .test) }}: template: :1:3: executing \"\" at <index .rmap (default .override .test)>: error calling index: index of untyped nil" interpolatedGenerator=nil
    --- FAIL: TestInterpolateGeneratorError/Error_templating (3.23s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x103481060]
```

**Expected behavior**

An invalid GoTemplate shouldn't crash the applicationset controller, it should be reflected in the status of the applicationset. 

**Screenshots**

<!-- If applicable, add screenshots to help explain your problem. -->

**Version**

```shell
❯ argocd version
argocd: v2.8.2+dbdfc71.dirty
  BuildDate: 2023-08-25T03:59:19Z
  GitCommit: dbdfc712702ce2f781910a795d2e5385a4f5a0f9
  GitTreeState: dirty
  GoVersion: go1.21.0
  Compiler: gc
  Platform: darwin/arm64
argocd-server: v2.8.2+dbdfc71
  BuildDate: 2023-08-24T20:05:39Z
  GitCommit: dbdfc712702ce2f781910a795d2e5385a4f5a0f9
  GitTreeState: clean
  GoVersion: go1.20.6
  Compiler: gc
  Platform: linux/amd64
  Kustomize Version: v5.0.1 2023-03-14T01:32:25Z
  Helm Version: v3.12.1+gf32a527
  Kubectl Version: v0.24.2
  Jsonnet Version: v0.20.0
```

**Logs**

```
ERRO[0005] error interpolating generator with other generator's parameter  error="failed to replace parameters in generator: failed to execute go template {{ index .registry_map (default .registry_override .jurisdiction) }}: template: :1:3: executing \"\" at <index .registry_map (default .registry_override .jurisdiction)>: error calling index: index of untyped nil" interpolatedGenerator=nil
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x107cf9960]

goroutine 332 [running]:
github.com/argoproj/argo-cd/v2/applicationset/generators.InterpolateGenerator(0x140006c0140, 0x140003c5380, 0x1, {0x0, 0x0, 0x0})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/generators/generator_spec_processor.go:156 +0x130
github.com/argoproj/argo-cd/v2/applicationset/generators.Transform({_, _, _, _, _, _, _, _, _, _}, ...)
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/generators/generator_spec_processor.go:56 +0x508
github.com/argoproj/argo-cd/v2/applicationset/generators.(*MatrixGenerator).getParams(0x140006c2078, {0x0, 0x0, 0x140013398c0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...}, ...)
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/generators/matrix.go:106 +0x56c
github.com/argoproj/argo-cd/v2/applicationset/generators.(*MatrixGenerator).GenerateParams(0x140006c2078, 0x140006586e0, 0x14000242000)
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/generators/matrix.go:56 +0x408
github.com/argoproj/argo-cd/v2/applicationset/generators.Transform({_, _, _, _, _, _, _, _, _, _}, ...)
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/generators/generator_spec_processor.go:67 +0x6b4
github.com/argoproj/argo-cd/v2/applicationset/controllers.(*ApplicationSetReconciler).generateApplications(_, {{{0x1093b0695, 0xe}, {0x14000a88a68, 0x14}}, {{0x14000a8a520, 0x1b}, {0x0, 0x0}, {0x14000665f46, ...}, ...}, ...})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/controllers/applicationset_controller.go:506 +0x204
github.com/argoproj/argo-cd/v2/applicationset/controllers.(*ApplicationSetReconciler).Reconcile(0x14000a9a4e0, {0x109ed8bf8, 0x14000fdd920}, {{{0x14000665f46, 0x6}, {0x14000a8a520, 0x1b}}})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/applicationset/controllers/applicationset_controller.go:116 +0x2f4
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x140011d0000, {0x109ed8bf8, 0x14000fdd920}, {{{0x14000665f46, 0x6}, {0x14000a8a520, 0x1b}}})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:114 +0x2bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0x140011d0000, {0x109ed8bf8, 0x14000fdd8f0}, {0x10997f5e0, 0x14000f44a60})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:311 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0x140011d0000, {0x109ed8b50, 0x140006b7c80})
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:266 +0x2bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:227 +0xa8
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
        /Users/rumstead/.gvm/pkgsets/go1.19.9/global/argo-cd/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:223 +0x3ec
Exiting.

```

--- a/applicationset/generators/generator_spec_processor.go
+++ b/applicationset/generators/generator_spec_processor.go
@@ -153,7 +153,7 @@ func InterpolateGenerator(requestedGenerator *argoprojiov1alpha1.ApplicationSetG
 	interpolatedGenerator, err := render.RenderGeneratorParams(requestedGenerator, params, useGoTemplate, goTemplateOptions)
 	if err != nil {
 		log.WithError(err).WithField("interpolatedGenerator", interpolatedGenerator).Error("error interpolating generator with other generator's parameter")
-		return *interpolatedGenerator, err
+		return argoprojiov1alpha1.ApplicationSetGenerator{}, err
 	}
 
 	return *interpolatedGenerator, nil

--- a/applicationset/generators/generator_spec_processor_test.go
+++ b/applicationset/generators/generator_spec_processor_test.go
@@ -515,7 +515,6 @@ func TestInterpolateGeneratorError(t *testing.T) {
 		want           argov1alpha1.ApplicationSetGenerator
 		expectedErrStr string
 	}{
-		//{{ index .registry_map (default .registry_override .jurisdiction) }}
 		{name: "Empty Gen", args: args{
 			requestedGenerator: nil,
 			params:             nil,

--- a/applicationset/generators/generator_spec_processor_test.go
+++ b/applicationset/generators/generator_spec_processor_test.go
@@ -501,3 +501,61 @@ func TestInterpolateGenerator_go(t *testing.T) {
 	assert.Equal(t, "production_01/west", interpolatedGenerator.Git.Files[0].Path)
 	assert.Equal(t, "https://production-01.example.com", interpolatedGenerator.Git.Files[1].Path)
 }
+
+func TestInterpolateGeneratorError(t *testing.T) {
+	type args struct {
+		requestedGenerator *argov1alpha1.ApplicationSetGenerator
+		params             map[string]interface{}
+		useGoTemplate      bool
+		goTemplateOptions  []string
+	}
+	tests := []struct {
+		name           string
+		args           args
+		want           argov1alpha1.ApplicationSetGenerator
+		expectedErrStr string
+	}{
+		//{{ index .registry_map (default .registry_override .jurisdiction) }}
+		{name: "Empty Gen", args: args{
+			requestedGenerator: nil,
+			params:             nil,
+			useGoTemplate:      false,
+			goTemplateOptions:  nil,
+		}, want: argov1alpha1.ApplicationSetGenerator{}, expectedErrStr: "generator is empty"},
+		{name: "No Params", args: args{
+			requestedGenerator: &argov1alpha1.ApplicationSetGenerator{},
+			params:             map[string]interface{}{},
+			useGoTemplate:      false,
+			goTemplateOptions:  nil,
+		}, want: argov1alpha1.ApplicationSetGenerator{}, expectedErrStr: ""},
+		{name: "Error templating", args: args{
+			requestedGenerator: &argov1alpha1.ApplicationSetGenerator{Git: &argov1alpha1.GitGenerator{
+				RepoURL:  "foo",
+				Files:    []argov1alpha1.GitFileGeneratorItem{{Path: "bar/"}},
+				Revision: "main",
+				Values: map[string]string{
+					"git_test":  "{{ toPrettyJson . }}",
+					"selection": "{{ default .override .test }}",
+					"resolved":  "{{ index .rmap (default .override .test) }}",
+				},
+			}},
+			params: map[string]interface{}{
+				"name":     "in-cluster",
+				"override": "foo",
+			},
+			useGoTemplate:     true,
+			goTemplateOptions: []string{},
+		}, want: argov1alpha1.ApplicationSetGenerator{}, expectedErrStr: "failed to replace parameters in generator: failed to execute go template {{ index .rmap (default .override .test) }}: template: :1:3: executing \"\" at <index .rmap (default .override .test)>: error calling index: index of untyped nil"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := InterpolateGenerator(tt.args.requestedGenerator, tt.args.params, tt.args.useGoTemplate, tt.args.goTemplateOptions)
+			if tt.expectedErrStr != "" {
+				assert.EqualError(t, err, tt.expectedErrStr)
+			} else {
+				require.NoError(t, err)
+			}
+			assert.Equalf(t, tt.want, got, "InterpolateGenerator(%v, %v, %v, %v)", tt.args.requestedGenerator, tt.args.params, tt.args.useGoTemplate, tt.args.goTemplateOptions)
+		})
+	}
+}

